Rubberduck.VBEEditor/SafeComWrappers/SafeEventedComWrapper.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.ComTypes;
+using Rubberduck.VBEditor.SafeComWrappers.Abstract;
+
+namespace Rubberduck.VBEditor.SafeComWrappers
+{
+    public abstract class SafeEventedComWrapper<TSource, TEventInterface> : SafeComWrapper<TSource>, ISafeEventedComWrapper
+        where TSource : class
+        where TEventInterface : class
+    {
+        private const int NotAdvising = -1;
+        private readonly object _lock = new object();
+        private IConnectionPoint _icp; // The connection point
+        private int _cookie = NotAdvising;     // The cookie for the connection
+
+        protected SafeEventedComWrapper(TSource target, bool rewrapping = false) : base(target, rewrapping)
+        {
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            DetachEvents();
+            base.Dispose(disposing);
+        }
+
+        public void AttachEvents()
+        {
+            lock (_lock)
+            {
+                if (IsWrappingNullReference)
+                {
+                    return;
+                }
+
+                if (_cookie != NotAdvising)
+                {
+                    return;
+                }
+
+                // Call QueryInterface for IConnectionPointContainer
+                var icpc = (IConnectionPointContainer) Target;
+
+                // Find the connection point for the source interface
+                var g = typeof(TEventInterface).GUID;
+                icpc.FindConnectionPoint(ref g, out _icp);
+
+                // Pass a pointer to the host to the connection point
+                _icp.Advise(this as TEventInterface, out _cookie);
```

Comment:
I would prefer to see this as a proper cast to disallow the null passthrough which shouldn't be needed?

Reply:
Not sure I follow. The `SafeEventedComWrapper` would require the implementation to have `TEventInterface` defined and thus can be cast to that interface. Unfortunately, there is no semantic means in C# to express that a generic type should be implemented by the class so I cannot enforce that the implementing class _must_ implement the event interface but I can at least hint it should be since the interface is required by the `TEventInterface` argument of the abstract class.

Reply:
I mean by changing to (TEventInterface)this